pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,214 @@
+package vitessshard
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+)
+
+func (r *ReconcileVitessShard) reconcileRollout(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
+	if !rollout.Scheduled(vts) {
+		// If the shard is not scheduled for a rolling update, silently bail out and do nothing.
+		return resultBuilder.Result()
+	}
+
+	for _, tablet := range vts.Status.Tablets {
+		if tablet.Ready != corev1.ConditionTrue {
+			if tablet.Ready != corev1.ConditionTrue {
+				// If all tablets aren't healthy, we should bail and not perform a rolling restart.
+				r.recorder.Eventf(vts, corev1.EventTypeWarning, "RollingRestartFailed", "all tablets are not healthy")
+				return resultBuilder.Result()
+			}
+		}
+	}
```

Comment:
In this same loop, we should also similarly bail out if any tablet has already been released. We have to wait until nothing is released before we decide to release anything else. I guess that means we'll have to list the pods before this loop and put them in a map by alias string.